overlord/snapstate/handlers.go

```diff
@@ -1042,12 +1090,26 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 	// record type
 	snapst.SetType(newInfo.GetType())
 
+	pb := NewTaskProgressAdapterLocked(t)
+
 	// XXX: this block is slightly ugly, find a pattern when we have more examples
 	model, _ := ModelFromTask(t)
 	err = m.backend.LinkSnap(newInfo, model, perfTimings)
 	if err != nil {
-		pb := NewTaskProgressAdapterLocked(t)
-		err := m.backend.UnlinkSnap(newInfo, pb)
+		var firstErr error
+		var err error
+		if len(snapst.Sequence) > 0 {
+			// save the disabled services for later
+			firstErr = m.saveSnapDisabledServices(snapst, pb)
```

Comment:
I don't think this should be done here. When link fails, service units will be removed too, so querying the status of disabled services won't produce  meaningful results.

Reply:
Okay, yes I think you're right. I had thought there would be a case where there would still be disabled services, but there would only be lost state if we did the following:

- the list of services in snapst was modified (it's not modified when we run LinkSnap, only after we do a successful restore-disabled-services, which itself is only done after a successful link-snap)
- any user code from snap hooks could have modified snap service state (no hooks are run in between unlink-snap and LinkSnap)

Reply:
Done